pandas/core/indexing.py

```diff
@@ -667,6 +667,8 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None):
                 if k not in self.obj:
                     if value is None:
                         self.obj[k] = np.nan
+                    elif is_array_like(value) and value.ndim == 2:
+                        self.obj[k] = value[:, i]
```

Comment:
is this depending on having `len(self.obj) == len(value)`?

Reply:
No this validation is done in this setitem

Reply:
comment here pointing back to this PR, can be follow-up

Reply:
Done